File: packages/scene-gl/src/standardPbrGlMeshMaterialRenderer.ts
import { getCameraViewProjectionMatrix4 } from '@flighthq/camera';
import { createMatrix4, getMatrix4Position, inverseMatrix4 } from '@flighthq/geometry';
import type { LinearColor } from '@flighthq/materials';
import { unpackColorToLinear } from '@flighthq/materials';
import { bindGlTexture } from '@flighthq/render-gl';
import type {
  Camera,
  GlMeshMaterialRenderer,
  GlRenderState,
  Material,
  MeshGeometry,
  SceneLightBlock,
  SceneRenderProxy,
  StandardPbrMaterial,
} from '@flighthq/types';

import { ensureGlMeshUpload } from './glMeshUpload';
import type { GlPbrDefineKey } from './glPbrPrelude';
import type { GlPbrProgram } from './glPbrProgramCache';
import { ensureGlPbrProgram } from './glPbrProgramCache';
import { getGlSceneRuntime } from './glSceneRuntime';

// The built-in StandardPbr forward-lit mesh-material renderer (GlMeshMaterialRenderer for
// StandardPbrMaterialKind). bind selects the uber-shader variant for the material's maps/alpha mode,
// uploads the shared per-run uniforms (camera view-projection + position, the packed light block),
// and the material's scalar/color uniforms and textures. draw uploads the geometry's GPU buffers
// lazily (cached by geometry.version), sets the per-draw model + normal matrices from the proxy, and
// issues the indexed draw over the proxy's subset with depth-test LESS + depth-write on and back-face
// culling unless the material is double-sided. See registerStandardPbrGlMaterial to install it.
export const standardPbrGlMeshMaterialRenderer: GlMeshMaterialRenderer = {
  bind(
    state: GlRenderState,
    material: Readonly<Material> | null,
    lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera>,
  ): void {
    const gl = state.gl;
    const pbr = material as Readonly<StandardPbrMaterial> | null;
    const program = ensureGlPbrProgram(state, defineKeyForMaterial(pbr));
    getGlSceneRuntime(state).activePbrProgram = program;

    gl.useProgram(program.program);

    // The render-effect pipeline owns enabling depth + binding the rgba16f scene target; this
    // renderer only fixes the test function/write and per-material cull. Depth state is set here so
    // a caller invoking drawScene without the full pipeline still gets correct occlusion.
    gl.enable(gl.DEPTH_TEST);
    gl.depthFunc(gl.LESS);
    gl.depthMask(true);

    if (pbr !== null && pbr.doubleSided) {
      gl.disable(gl.CULL_FACE);
    } else {
      gl.enable(gl.CULL_FACE);
      gl.cullFace(gl.BACK);
    }

    bindGlPbrCamera(gl, program, camera);
    bindGlPbrLights(gl, program, lights);
    bindGlPbrMaterialUniforms(state, program, pbr);
  },

  draw(state: GlRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    const gl = state.gl;
    const program = getGlSceneRuntime(state).activePbrProgram;
    if (program === null) return;

    gl.uniformMatrix4fv(program.locModel, false, proxy.worldMatrix.m);
    gl.uniformMatrix3fv(program.locNormalMatrix, false, proxy.normalMatrix.m);

    const upload = ensureGlMeshUpload(state, geometry);
    const subset = proxy.subset;

    if (upload.indexBuffer !== null) {
      const elementSize = upload.indexType === gl.UNSIGNED_INT ? 4 : 2;
      gl.drawElements(gl.TRIANGLES, subset.indexCount, upload.indexType, subset.indexOffset * elementSize);
    } else {
      gl.drawArrays(gl.TRIANGLES, subset.indexOffset, subset.indexCount);
    }
  },
};

// The feature define key for a StandardPbr material: which optional maps are present and whether
// alpha-mask cutoff is active. Drives both the program-cache variant and the bound textures.
function defineKeyForMaterial(material: Readonly<StandardPbrMaterial> | null): GlPbrDefineKey {
  return {
    alphaMaskEnabled: material !== null && material.alphaMode === 'mask',
    hasBaseColorMap: material !== null && material.baseColorMap !== null && material.baseColorMap.image !== null,
    hasNormalMap: material !== null && material.normalMap !== null && material.normalMap.image !== null,
  };
}

function bindGlPbrCamera(gl: WebGL2RenderingContext, program: Readonly<GlPbrProgram>, camera: Readonly<Camera>): void {
  const aspect = camera.projection.kind === 'perspective' ? camera.projection.aspect : 1;
  getCameraViewProjectionMatrix4(scratchViewProjection, camera, aspect !== 0 ? aspect : 1);
  gl.uniformMatrix4fv(program.locViewProjection, false, scratchViewProjection.m);

  // Camera world position = translation of the inverse view matrix (view is world->view).
  inverseMatrix4(scratchInverseView, camera.view);
  getMatrix4Position(scratchCameraPosition, scratchInverseView);
  gl.uniform3f(program.locCameraPosition, scratchCameraPosition.x, scratchCameraPosition.y, scratchCameraPosition.z);
}

// Uploads the packed light block to the directional/ambient uniforms. The block layout (std140)
// matches SceneLightBlock.data: directional { direction.xyz @0, _pad, radiance.rgb @4, _pad } then
// ambient { radiance.rgb @8, _pad }. Radiance is already linear, premultiplied by intensity.
function bindGlPbrLights(
  gl: WebGL2RenderingContext,
  program: Readonly<GlPbrProgram>,
  lights: Readonly<SceneLightBlock>,
): void {
  const data = lights.data;
  gl.uniform4f(program.locDirectional, data[0], data[1], data[2], 0);
  gl.uniform4f(program.locDirectionalRadiance, data[4], data[5], data[6], 0);
  gl.uniform3f(program.locAmbientRadiance, data[8], data[9], data[10]);
  gl.uniform1f(program.locDirectionalCount, lights.directionalCount);
  gl.uniform1f(program.locAmbientCount, lights.ambientCount);
}

function bindGlPbrMaterialUniforms(
  state: GlRenderState,
  program: Readonly<GlPbrProgram>,
  material: Readonly<StandardPbrMaterial> | null,
): void {
  const gl = state.gl;
  if (material === null) {
    gl.uniform4f(program.locBaseColor, 1, 1, 1, 1);
    gl.uniform1f(program.locMetallic, 0);
    gl.uniform1f(program.locRoughness, 1);
    gl.uniform1f(program.locNormalScale, 1);
    gl.uniform3f(program.locEmissive, 0, 0, 0);
    gl.uniform1f(program.locEmissiveStrength, 1);
    gl.uniform1f(program.locAlphaCutoff, 0.5);
    return;
  }

  unpackColorToLinear(scratchRgba, material.baseColor);
  gl.uniform4f(program.locBaseColor, scratchRgba[0], scratchRgba[1], scratchRgba[2], scratchRgba[3]);
  gl.uniform1f(program.locMetallic, material.metallic);
  gl.uniform1f(program.locRoughness, material.roughness);
  gl.uniform1f(program.locNormalScale, material.normalScale);

  unpackColorToLinear(scratchRgba, material.emissive);
  gl.uniform3f(program.locEmissive, scratchRgba[0], scratchRgba[1], scratchRgba[2]);
  gl.uniform1f(program.locEmissiveStrength, material.emissiveStrength);
  gl.uniform1f(program.locAlphaCutoff, material.alphaCutoff);

  const baseColorMap = material.baseColorMap;
  if (baseColorMap !== null && baseColorMap.image !== null && baseColorMap.image.source !== null) {
    gl.activeTexture(gl.TEXTURE0);
    bindGlTexture(state, baseColorMap.image.source);
    gl.uniform1i(program.locBaseColorMap, 0);
  }

  const normalMap = material.normalMap;
  if (normalMap !== null && normalMap.image !== null && normalMap.image.source !== null) {
    gl.activeTexture(gl.TEXTURE1);
    bindGlTexture(state, normalMap.image.source);
    gl.uniform1i(program.locNormalMap, 1);
  }
}

const scratchViewProjection = createMatrix4();
const scratchInverseView = createMatrix4();
const scratchCameraPosition = { x: 0, y: 0, z: 0 };
const scratchRgba: LinearColor = [0, 0, 0, 0];

File: crates/flighthq-scene-gl/src/standard_pbr_gl_mesh_material_renderer.rs
//! The built-in StandardPbr forward-lit mesh-material renderer.
//!
//! Ports `@flighthq/scene-gl` `standardPbrGlMeshMaterialRenderer.ts`. `bind`
//! selects the uber-shader variant for the material's maps/alpha mode, uploads the
//! shared per-run uniforms (camera view-projection + position, the packed light
//! block) via [`gl_pbr_standard_bind`](crate::gl_pbr_standard_bind), and the
//! material's `standard` block through the shared
//! [`bind_gl_pbr_standard_block`]. `draw` sets the per-draw model + normal matrices
//! and issues the indexed draw over the proxy's subset with depth-test LESS +
//! depth-write on and back-face culling unless the material is double-sided.

use flighthq_render_gl::GlRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::pbr_material::{MaterialAlphaMode, StandardPbrMaterial};
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};
use glow::HasContext;

use crate::gl_mesh_material_registry::{GlMeshMaterialRenderer, MeshMaterial};
use crate::gl_pbr_prelude::GlPbrDefineKey;
use crate::gl_pbr_standard_bind::{bind_gl_pbr_mesh_common, build_gl_pbr_standard_define_key};
use crate::gl_scene_runtime::{GlMeshUpload, GlSceneRuntime};

/// The built-in StandardPbr forward-lit mesh-material renderer
/// (`GlMeshMaterialRenderer` for the StandardPbr material kind). See
/// [`register_standard_pbr_gl_material`](crate::register_standard_pbr_gl_material)
/// to install it.
pub struct StandardPbrGlMeshMaterialRenderer;

impl GlMeshMaterialRenderer for StandardPbrGlMeshMaterialRenderer {
    fn bind(
        &self,
        state: &mut GlRenderState,
        scene: &mut GlSceneRuntime,
        material: Option<&dyn MeshMaterial>,
        lights: &SceneLightBlock,
        camera: &Camera,
    ) {
        let pbr = material.and_then(|m| m.as_standard_pbr());
        let standard = pbr.map(|p| &p.standard);
        let alpha_mask_enabled = pbr.is_some_and(|p| p.alpha_mode == MaterialAlphaMode::Mask);

        // StandardPbr sets no extension flag — the base variant.
        let key = build_gl_pbr_standard_define_key(standard, alpha_mask_enabled);
        bind_gl_pbr_mesh_common(
            state,
            scene,
            &key,
            pbr.is_some_and(|p| p.double_sided),
            standard,
            alpha_cutoff(pbr),
            lights,
            camera,
        );
    }

    fn draw(
        &self,
        state: &mut GlRenderState,
        scene: &mut GlSceneRuntime,
        proxy: &SceneRenderProxy,
        upload: &GlMeshUpload,
    ) {
        draw_gl_pbr_mesh_subset(state, scene, proxy, upload);
    }
}

/// Sets the per-draw model + normal matrices on the active PBR program and issues
/// the indexed (or array) draw over the proxy's subset. The shared draw body for
/// StandardPbr and every extension renderer — the lobe uniforms differ at `bind`,
/// but the geometry draw is identical (ports the shared TS `drawGlMeshSubset`). A
/// no-op when no PBR program is active (bind was skipped).
pub fn draw_gl_pbr_mesh_subset(
    state: &mut GlRenderState,
    scene: &mut GlSceneRuntime,
    proxy: &SceneRenderProxy,
    upload: &GlMeshUpload,
) {
    let Some(program) = scene.active_pbr_program.clone() else {
        return;
    };
    let gl = &state.gl;
    unsafe {
        gl.uniform_matrix_4_f32_slice(program.loc_model.as_ref(), false, &proxy.world_matrix.m);
        gl.uniform_matrix_3_f32_slice(
            program.loc_normal_matrix.as_ref(),
            false,
            &proxy.normal_matrix.m,
        );

        let subset = proxy.subset;
        if upload.index_buffer.is_some() {
            let element_size = if upload.index_type == glow::UNSIGNED_INT {
                4
            } else {
                2
            };
            gl.draw_elements(
                glow::TRIANGLES,
                subset.index_count as i32,
                upload.index_type,
                (subset.index_offset * element_size) as i32,
            );
        } else {
            gl.draw_arrays(
                glow::TRIANGLES,
                subset.index_offset as i32,
                subset.index_count as i32,
            );
        }
    }
}

/// The alpha-cutoff to upload: the material's own cutoff, or the neutral `0.5`
/// default when no concrete material resolved. Mirrors the TS trailing
/// `gl.uniform1f(program.locAlphaCutoff, …)`.
fn alpha_cutoff(pbr: Option<&StandardPbrMaterial>) -> f32 {
    pbr.map_or(0.5, |p| p.alpha_cutoff)
}

/// The feature define key for the bound material: the alpha-mask flag plus the
/// standard-block map flags from the concrete `StandardPbrMaterial`. Re-exported for
/// tests and callers that build a StandardPbr key without a live GL context.
pub fn build_standard_pbr_define_key(
    pbr: Option<&StandardPbrMaterial>,
    alpha_mask_enabled: bool,
) -> GlPbrDefineKey {
    build_gl_pbr_standard_define_key(pbr.map(|p| &p.standard), alpha_mask_enabled)
}

#[cfg(test)]
mod tests {
    use super::*;

    // The GL bind/draw path requires a live glow context (a real GPU/display), so
    // it is validated functionally (the parity matrix at the `gl` cell), not by a
    // unit test — matching `flighthq-render-gl`'s no-device test posture. The pure
    // material-read helpers are assertion-ported here.

    // build_standard_pbr_define_key

    #[test]
    fn build_standard_pbr_define_key_defaults_to_no_maps_no_alpha_mask() {
        let key = build_standard_pbr_define_key(None, false);
        assert_eq!(key, GlPbrDefineKey::default());
    }

    #[test]
    fn build_standard_pbr_define_key_reads_no_maps_from_the_default_material() {
        let material = StandardPbrMaterial::default();
        let key = build_standard_pbr_define_key(Some(&material), false);
        assert!(!key.has_base_color_map);
        assert!(!key.has_normal_map);
        assert!(!key.alpha_mask_enabled);
        assert!(!key.specular_enabled);
    }

    #[test]
    fn build_standard_pbr_define_key_enables_alpha_mask() {
        let material = StandardPbrMaterial::default();
        let key = build_standard_pbr_define_key(Some(&material), true);
        assert!(key.alpha_mask_enabled);
    }

    // alpha_cutoff

    #[test]
    fn alpha_cutoff_falls_back_to_the_neutral_default() {
        assert_eq!(alpha_cutoff(None), 0.5);
    }
}
